File: kl/codegen.go
package kl

import (
	"fmt"
	"os"
)

func GenerateC(args ...Obj) Obj {
	filePath := mustString(args[0])
	bc := ListToSlice(args[1])

	out, err := os.Create(filePath)
	if err != nil {
		return MakeError(fmt.Sprintf("open file fail %s", err))
	}
	defer out.Close()

	// Generate "#include ..."
	fmt.Fprintf(out, "#include \"runtime.h\"\n")
	fmt.Fprintf(out, "#include <stdlib.h>\n\n")

	labelSym := MakeSymbol("label")
	// Generate declare...
	for _, fn := range bc {
		if Car(fn) == labelSym {
			fmt.Fprintf(out, "static void %s (struct controlFlow *ctx);\n", symbolString(Cadr(fn)))
		} else {
			fmt.Println("???  = ", ObjString(Car(fn)))
		}
	}
	fmt.Fprintf(out, "nativeFuncPtr export = %s;\n", symbolString(Cadr(bc[0])))

	fmt.Fprintf(out, "\n\n")

	// Generate each function...
	for _, fn := range bc {
		if err := generateFunc(out, fn); err != nil {
			fmt.Println("err =", err)
			return MakeError(err.Error())
		}
	}
	return True
}

func symbolString(sym Obj) string {
	str := GetSymbol(sym)
	if str[0] == '#' {
		return str[1:]
	}
	return str
}

func generateInst(w *os.File, inst Obj) error {
	kind := GetSymbol(Car(inst))
	switch kind {
	case "mov":
		// (mov SRC DST)
		src := Cadr(inst)
		dst := Car(Cdr(Cdr(inst)))
		fmt.Fprintf(w, "%s = %s;\n", symbolString(dst), symbolString(src))
	case "%const":
		// (const Number DST)
		// (const () DST)
		// (cons "xxx" DST)
		dst := Car(Cdr(Cdr(inst)))
		c := Cadr(inst)
		switch {
		case IsNumber(c):
			fmt.Fprintf(w, "Obj %s = makeNumber(%d);\n", symbolString(dst), GetInteger(Cadr(inst)))
		case IsString(c):
			str := GetString(c)
			fmt.Fprintf(w, "Obj %s = makeString(\"%s\", %d);\n", symbolString(dst), str, len(str))
		case IsSymbol(c):
			fmt.Fprintf(w, "Obj %s = intern(\"%s\");\n", symbolString(dst), symbolString(c))
		case c == Nil:
			fmt.Fprintf(w, "Obj %s = Nil;\n", symbolString(dst))
		case c == True:
			fmt.Fprintf(w, "Obj %s = True;\n", symbolString(dst))
		case c == False:
			fmt.Fprintf(w, "Obj %s = False;\n", symbolString(dst))
		}
	case "%closure-ref":
		dst := Cadr(inst)
		clo := Cadr(Cdr(inst))
		idx := Cadr(Cdr(Cdr(inst)))
		fmt.Fprintf(w, "Obj %s = closureRef(%s, %d);\n", symbolString(dst), symbolString(clo), GetInteger(idx))
	case "%closure":
		// (closure DST FuncName Required FV0 FV1 ...)
		tmp := ListToSlice(Cdr(inst))
		dst := tmp[0]
		funcName := tmp[1]
		required := tmp[2]
		captured := tmp[3:]
		args := fmt.Sprintf("%s , %d, %d", symbolString(funcName),
			GetInteger(required),
			len(captured))
		for i := 0; i < len(captured); i++ {
			args += ", "
			args = args + symbolString(captured[i])
		}
		fmt.Fprintf(w, "Obj %s = makeNative(%s);\n", symbolString(dst), args)
	case "%builtin":
		// (builtin OP (ARG1 ARG2 ...) DST)
		op := Cadr(inst)
		args := Cadr(Cdr(inst))
		dst := Car(Cdr(Cdr(Cdr(inst))))
		generateCBuiltinCall(w, op, args, dst)
	case "%return":
		fmt.Fprintf(w, "return ctxReturn(ctx, %s);\n", symbolString(Cadr(inst)))
	case "%call-def":
		fn := Cadr(inst)
		args := ListToSlice(Cadr(Cdr(inst)))
		dst := Car(Cdr(Cdr(Cdr(inst))))
		fmt.Fprintf(w, `Obj %s = Call(intern("%s"), %d`,
			symbolString(dst),
			symbolString(fn),
			len(args))
		for _, arg := range args {
			fmt.Fprintf(w, ", ")
			fmt.Fprintf(w, symbolString(arg))
		}
		fmt.Fprintf(w, ");\n")
	case "if":
		// (if a b c)
		a := Cadr(inst)
		b := Car(Cdr(Cdr(inst)))
		c := Car(Cdr(Cdr(Cdr(inst))))
		if err := generateCIfExpr(w, a, b, c); err != nil {
			return err
		}
	default:
		return fmt.Errorf("unknown instruct: %s\n", kind)
	}
	return nil
}

func generateInsts(w *os.File, sexp Obj) error {
	instructs := ListToSlice(sexp)
	for _, inst := range instructs {
		if err := generateInst(w, inst); err != nil {
			return err
		}
	}
	return nil
}

func generateFunc(w *os.File, sexp Obj) error {
	if symbolString(Car(sexp)) != "label" {
		return fmt.Errorf("invalid input:%s\n", ObjString(sexp))
	}
	// (label FuncName (Args ...) Body)
	funcName := symbolString(Cadr(sexp))
	fmt.Fprintf(w, "static void %s(struct controlFlow *ctx) {\n", funcName)

	args := ListToSlice(Cadr(Cdr(sexp)))
	for ith, arg := range args {
		fmt.Fprintf(w, "Obj %s = ctxGet(ctx, %d);\n", symbolString(arg), ith)
	}

	if err := generateInsts(w, Cdr(Cdr(Cdr(sexp)))); err != nil {
		return err
	}
	fmt.Fprintf(w, "}\n\n")
	return nil
}

func generateCIfExpr(w *os.File, a, b, c Obj) error {
	fmt.Fprintf(w, "if (%s == True) {\n", symbolString(a))
	if err := generateInsts(w, b); err != nil {
		return err
	}
	fmt.Fprintf(w, "} else {\n")
	if err := generateInsts(w, c); err != nil {
		return err
	}
	fmt.Fprintf(w, "}\n")
	return nil
}

func generateCBuiltinCall(w *os.File, op, args, dst Obj) {
	input := ListToSlice(args)
	fmt.Fprintf(w, `Obj %s = Call(intern("%v"), %d`, symbolString(dst),
		symbolString(op), len(input))
	for _, arg := range input {
		fmt.Fprintf(w, ", %s", symbolString(arg))
	}
	fmt.Fprintf(w, ");\n")
}

File: src/codegen.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "runtime.h"

static int genIfExpr(FILE *w, Obj a, Obj b, Obj c);

static int
listLen(Obj l) {
  int count = 0;
  while(l != Nil) {
    count++;
    l = cdr(l);
  }
  return count;
}

static void
listToSlice(Obj* res, int size, Obj l) {
  for (int i=0; i<size; i++) {
    res[i] = car(l);
    l = cdr(l);
  }
  assert(l == Nil);
  return;
}

static const char*
symbolString(Obj sym) {
 const char *str = symbolStr(sym);
 if (str[0] == '#') {
   return str+1;
 }
 return str;
}

static Obj symLet, symConst, symClosure, symReturn, symBuiltin, symCall, symClosureRef, symTailCall, symGlobal;

static void
init() {
  symLet = intern("let");
  symConst = intern("%const");
  symClosureRef = intern("%closure-ref");
  symClosure = intern("%closure");
  symBuiltin = intern("%builtin");
  symGlobal = intern("%global");
  symReturn = intern("%return");
  symCall = intern("%call");
  symTailCall = intern("%tailcall");
}

static int
genConst(FILE *w, Obj inst) {
  // (const Number)
  // (const ())
  // (cons "xxx")
  Obj c = cadr(inst);
  if  (isfixnum(c)) {
    fprintf(w, "makeNumber(%ld)", fixnum(cadr(inst)));
  } else if (isstring(c)) {
    char *str = stringStr(c);
    fprintf(w, "makeString(\"%s\", %ld)", str, strlen(str));
  } else if (issymbol(c)) {
    fprintf(w, "intern(\"%s\")", symbolString(c));
  } else if (c == Nil) {
    fprintf(w, "Nil");
  } else if (c == True) {
    fprintf(w, "True");
  } else if (c == False) {
    fprintf(w, "False");
  } else {
    fprintf(w, "unknown const");
    return -1;
  }
  return 0;
}

static int
genCall(FILE *w, Obj inst, bool tail) {
  // (%tailcall x y z ...)
  if (tail) {
    fprintf(w, "return ctxTailCall(");
  } else {
    fprintf(w, "call(");
  }

  bool first = true;
  for (Obj o = cdr(inst); o != Nil; o = cdr(o)) {
    if (first) { first = false; } else { fprintf(w, ", "); }
    fprintf(w, "%s", symbolString(car(o)));
  }
  fprintf(w, ")");

  if (tail) {
    fprintf(w, ";\n");
  }
  return 0;
}

static int
genInst(FILE *w, Obj inst) {
  Obj kind = car(inst);
  if (kind == symConst) {
    return genConst(w, inst);
  } else if (kind == symClosureRef) {
    // (%closure-ref #clo282 0)
    Obj clo = cadr(inst);
    Obj idx = cadr(cdr(inst));
    fprintf(w, "closureRef(%s, %ld)", symbolString(clo), fixnum(idx));
  } else if (kind == symClosure) {
    // (%closure FuncName Required FV0 FV1 ...)
    int len = listLen(cdr(inst));
    Obj tmp[len];
    listToSlice(tmp, len, cdr(inst));
    Obj funcName = tmp[0];
    Obj required = tmp[1];
    /* Obj* captured = tmp+2; */
    char buf[256];
    sprintf(buf, "%s , %ld, %d", symbolString(funcName),
                       fixnum(required),
                       len-2);
    /* for (int i = 0; i < len-3; i++) { */
    /* args += ", "; */
    /* args = args + symbolString(captured[i]); */
    /* } */
    fprintf(w, "makeNative(%s)", buf);
  } else if (kind == symBuiltin) {
    // (%builtin OP)
    fprintf(w, "symbolGet(intern(%s))", symbolString(cadr(inst)));
  } else if (kind == symGlobal) {
    // (%global OP)
    fprintf(w, "symbolGet(intern(%s))", symbolString(cadr(inst)));
  } else if (kind == symCall) {
    // (%call x y z ...)
    return genCall(w, inst, false);
  } else if (kind == symTailCall) {
    // (%tailcall x y z ...)
    return genCall(w, inst, true);
  } else if (kind == symIf) {
    // (if a b c)
    Obj a = cadr(inst);
    Obj b = caddr(inst);
    Obj c = car(cdr(cdr(cdr(inst))));
    return genIfExpr(w, a, b, c);
  } else if (kind == symReturn) {
    // (%return xx)
    fprintf(w, "return ctxReturn(ctx, %s);\n", symbolString(cadr(inst)));
  } else {
    printf("unknown instruct: %s\n", symbolStr(kind));
    return -1;
  }
  return 0;
}

static Obj
genLet(FILE *w, Obj inst) {
  Obj var = cadr(inst);
  Obj val = caddr(inst);
  Obj remain = caddr(cdr(inst));
  fprintf(w, "%s = ", symbolString(var));
  int res = genInst(w, val);
  if (res != 0) {
    return Undef;
  }
  fprintf(w, ";\n");
  return remain;
}

static int
genInsts(FILE* w, Obj sexp) {
  while(sexp != Undef) {
    if (car(sexp) != symLet) {
      return genInst(w, sexp);
    }
    sexp = genLet(w, sexp);
  }
  printf("something wrong");
  sexpWrite(NULL, sexp);
  printf("\n");
  return -1;
}

static int
genIfExpr(FILE *w, Obj a, Obj b, Obj c) {
	fprintf(w, "if (%s == True) {\n", symbolString(a));
	int err = genInsts(w, b);
  if (err != 0) {
    return err;
  }
	fprintf(w, "} else {\n");
	err = genInsts(w, c);
  if (err != 0) {
		return err;
	}
	fprintf(w, "}\n");
	return 0;
}

static int
genFunc(FILE *w, Obj sexp) {
  if (car(sexp) != intern("label")) {
    printf("invalid input:\n");
    sexpWrite(w, sexp);
    return -1;
  }
  // (label FuncName (Args ...) Body)
  const char *funcName = symbolString(cadr(sexp));
  fprintf(w, "static void %s(struct controlFlow *ctx) {\n", funcName);

  int len = listLen(cadr(cdr(sexp)));
  Obj args[len];
  listToSlice(args, len, cadr(cdr(sexp)));
  for (int i=0; i<len; i++) {
    Obj arg = args[i];
    fprintf(w, "Obj %s = ctxGet(ctx, %d);\n", symbolString(arg), i);
  }

  int err = genInsts(w, car(cdr(cdr(cdr(sexp)))));
  if (err != 0) {
    return err;
  }
  fprintf(w, "}\n\n");
  return 0;
}

void
builtinGenerateC(struct controlFlow* ctx) {
  Obj dst = ctxGet(ctx, 1);
  Obj bc = ctxGet(ctx, 2);
  assert(isstring(dst));
  char* outFileName = stringStr(dst);
  FILE* outFile = fopen(outFileName, "w");
  if (outFile == NULL) {
    printf("open output file %s failed", outFileName);
    goto exit0;
  }

  if (symLet == 0) {
    init();
  }

  int len = listLen(bc);
  Obj* tmp = malloc(len * sizeof(Obj));
  if (tmp == NULL) {
    goto exit1;
  }

  listToSlice(tmp, len, bc);

  for (int i=0; i<len; i++) {
    Obj label = tmp[i];
    int err = genFunc(outFile, label);
    if (err != 0) {
      break;
    }
  }

  free(tmp);
exit1:
  fclose(outFile);
exit0:
  return ctxReturn(ctx, dst);
}

void
builtinReadFileAsSexp(struct controlFlow* ctx) {
  Obj arg = ctxGet(ctx, 1);
  assert(isstring(arg));
  char* fileName = stringStr(arg);
  FILE* f = fopen(fileName, "r");
  if (f == NULL) {
    printf("open file fail %s\n", fileName);
    goto exit0;
  }
  Obj ast = sexpRead(f);
  fclose(f);
  ctxReturn(ctx, ast);

exit0:
  return;
}

int
main() {
  char *input = "fact.bc";
  char *output = "gen.c";

  coraInit();

  Obj bc = Call(intern("read-file-as-sexp"), 1, makeString(input, strlen(input)));
  Obj res = Call(intern("generate-c"), 2, makeString(output, strlen(output)), bc);
  return 0;
}
